applications/serial_lte_modem/doc/slm_description.rst

```diff
@@ -0,0 +1,501 @@
+
+.. include:: /includes/build_and_run_nrf9160.txt
+
+
+Testing
+=======
+
+The testing instructions focus on testing the application with a PC client.
+If you have an nRF52 Series DK running a client application, you can also use this DK for testing the different scenarios.
+
+|test_sample|
+
+1. |connect_kit|
+#. :ref:`Connect to the kit with LTE Link Monitor <lte_connect>`.
+   If you want to use a different terminal emulator, see `Connecting with a PC`_.
+#. Reset the kit.
+#. Observe that the nRF9160 DK sends a ``Ready\r\n`` message on UART.
+#. Send AT commands and observe the responses from the nRF9160 DK.
+   See the following sections for typical test cases.
+
+
+Testing generic AT commands
+---------------------------
+
+Complete the following steps to test the functionality provided by the :ref:`SLM_AT_gen`:
+
+1. Retrieve the version of the serial LTE modem application.
+
+   .. parsed-literal::
+      :class: highlight
+
+      **AT#XSLMVER**
+      #XSLMVER: 1.4
+      OK
+
+#. Read the current baud rate.
+
+   .. parsed-literal::
+      :class: highlight
+
+      **AT#XSLMUART?**
+      #SLMUART: 115200
+      OK
+
+   You can change the used baud rate with the corresponding set command, but note that LTE Link Monitor requires 115200 bps for communication.
+
+#. Retrieve a list of all supported proprietary AT commands.
+
+   .. parsed-literal::
+      :class: highlight
+
+      **AT#XCLAC**
+      AT#XSLMVER
+      AT#XSLEEP
+      AT#XCLAC
+      AT#XSOCKET
+      AT#XSOCKETOPT
+      AT#XBIND
+      *[...]*
+      OK
+
+#. Check the supported values for the sleep command, then put the board in sleep mode.
+
+   .. parsed-literal::
+      :class: highlight
+
+      **AT#XSLEEP=?**
+      #XSLEEP: (0, 1)
+      OK
+
+      **AT#XSLEEP=1**
+
+   Reset the board to exit sleep mode.
+   If you are testing with an external MCU and :option:`SLM_GPIO_WAKEUP` is enabled, you can wake up by GPIO as well.
+
+
+
+Testing TCP/IP AT commands
+--------------------------
+
+The following sections show how to test the functionality provided by the :ref:`SLM_AT_TCPIP`.
+
+TCP client
+~~~~~~~~~~
+
+1. Establish and test a TCP connection:
+
+   a. Check the available values for the XSOCKET command.
+
+      .. parsed-literal::
+        :class: highlight
+
+        **AT#XSOCKET=?**
+        #XSOCKET: (0, 1), (1, 2), (0, 1), <sec-tag>
+        OK
+
+   #. Open a TCP socket, read information (handle, protocol, and role) about the open socket, and set the receive time-out of the open socket to 30 seconds.
+
+      .. parsed-literal::
+        :class: highlight
+
+        **AT#XSOCKET=1,1,0**
+        #XSOCKET: 2, 1, 0, 6
+        OK
+
+        **AT#XSOCKET?**
+        #XSOCKET: 2, 6, 0
+        OK
+
+        **AT#XSOCKETOPT=1,20,30**
+        OK
+
+   #. Connect to a TCP server on a specified port.
+      Replace *example.com* with the host name or IPv4 address of a TCP server and *1234* with the corresponding port.
```

Comment:
Are there any default servers that can be used for testing? Or are there instructions for how to set one up?
Is there a specific port number (like 2442), or is this random?

Reply:
How about use famous server like Google? Below is an example to use TCP command to connect to google.com and send test request/receive response.

AT#XSOCKET=1,1,0
#XSOCKET: 1, 1, 0, 6
OK
AT#XCONNECT="google.com",80
#XCONNECT: 1
OK
AT#XSEND=1,"HEAD / HTTP/1.1"
#XSEND: 15
OK
AT#XSEND=0,"0D0A"
#XSEND: 2
OK
AT#XSEND=1,"Host: www.google.com:443"
#XSEND: 24
OK
AT#XSEND=0,"0D0A"
#XSEND: 2
OK
AT#XSEND=1,"Connection: close"
#XSEND: 17
OK
AT#XSEND=0,"0D0A0D0A"
#XSEND: 4
OK
AT#XRECV
HTTP/1.1 200 OK
Content-Type: text/html; charset=ISO-8859-1
P3P: CP="This is not a P3P policy! See g.co/p3phelp for more info."
Date: Tue, 06 Oct 2020 06:17:59 GMT
Server: gws
X-XSS-Protection: 0
X-Frame-Options: SAMEORIGIN
Transfer-Encoding: chunked
Expires: Tue, 06 Oct 2020 06:17:59 GMT
Cache-Control: private
Set-Cookie: 1P_JAR=2020-10-06-06; expires=Thu, 05-Nov-2020 06:17:59 GMT; path=/; domain=.google.com; Secure
Set-Cookie: NID=204=hWzzT1EYH_rqEFR-ErQHEkxl-rc3DMQolj51fG_fCdPusz09YzoMhNkDNRII6RSmWOzbyyREk9terY5fNQMBu6b8MJa1DB-AxmXKmCkwP6Uys3t3QH82qRH2o3t3
#XRECV: 1, 576
OK
AT#XRECV
hrBgFfJBImpzbp4jFIqZJ6On-BWSo5HZJYrVWFQyXUJL0J0; expires=Wed, 07-Apr-2021 06:17:59 GMT; path=/; domain=.google.com; HttpOnly
Connection: close


#XRECV: 1, 147
OK

Reply:
As for the port number, it depends on the layer above TCP/UDP. Here is a list of well-known ports
https://en.wikipedia.org/wiki/List_of_TCP_and_UDP_port_numbers

Reply:
This fails for me at:

AT#XCONNECT="google.com",80
ERROR

That's with both LTE Link Monitor and Putty. I might be doing something wrong of course, but I'm following the instructions ...

Reply:
We tested again Nordic's internal test server, "vmi36865.contabo.host".
We cannot recommend any other server as the TCP connection service there is unknown.

Reply:
So we should keep the instructions as they are, with "example.com" and port "1234"?

Reply:
why not, unless we have a free TCP server online